// @wyw-in-js/transform/types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.findIdentifiers = exports.TransformCacheCollection = exports.peek = exports.getVisitorKeys = exports.applyProcessors = exports.getFileIdx = exports.isNode = exports.EventEmitter = exports.withDefaultServices = exports.loadWywOptions = exports.syncResolveImports = exports.asyncResolveImports = exports.transformUrl = exports.Entrypoint = exports.prepareCode = exports.baseHandlers = exports.parseFile = exports.EvaluatedEntrypoint = exports.UnprocessedEntrypointError = exports.isUnprocessedEntrypointError = exports.transform = exports.shaker = exports.DefaultModuleImplementation = exports.Module = exports.withTransformMetadata = exports.getTransformMetadata = exports.preeval = exports.babelTransformPlugin = exports.createFileReporter = exports.slugify = void 0;
var shared_1 = require("@wyw-in-js/shared");
Object.defineProperty(exports, "slugify", { enumerable: true, get: function () { return shared_1.slugify; } });
var fileReporter_1 = require("./debug/fileReporter");
Object.defineProperty(exports, "createFileReporter", { enumerable: true, get: function () { return fileReporter_1.createFileReporter; } });
var babel_transform_1 = require("./plugins/babel-transform");
Object.defineProperty(exports, "babelTransformPlugin", { enumerable: true, get: function () { return __importDefault(babel_transform_1).default; } });
var preeval_1 = require("./plugins/preeval");
Object.defineProperty(exports, "preeval", { enumerable: true, get: function () { return __importDefault(preeval_1).default; } });
var TransformMetadata_1 = require("./utils/TransformMetadata");
Object.defineProperty(exports, "getTransformMetadata", { enumerable: true, get: function () { return TransformMetadata_1.getTransformMetadata; } });
Object.defineProperty(exports, "withTransformMetadata", { enumerable: true, get: function () { return TransformMetadata_1.withTransformMetadata; } });
var module_1 = require("./module");
Object.defineProperty(exports, "Module", { enumerable: true, get: function () { return module_1.Module; } });
Object.defineProperty(exports, "DefaultModuleImplementation", { enumerable: true, get: function () { return module_1.DefaultModuleImplementation; } });
var shaker_1 = require("./shaker");
Object.defineProperty(exports, "shaker", { enumerable: true, get: function () { return __importDefault(shaker_1).default; } });
var transform_1 = require("./transform");
Object.defineProperty(exports, "transform", { enumerable: true, get: function () { return transform_1.transform; } });
var UnprocessedEntrypointError_1 = require("./transform/actions/UnprocessedEntrypointError");
Object.defineProperty(exports, "isUnprocessedEntrypointError", { enumerable: true, get: function () { return UnprocessedEntrypointError_1.isUnprocessedEntrypointError; } });
Object.defineProperty(exports, "UnprocessedEntrypointError", { enumerable: true, get: function () { return UnprocessedEntrypointError_1.UnprocessedEntrypointError; } });
__exportStar(require("./types"), exports);
var EvaluatedEntrypoint_1 = require("./transform/EvaluatedEntrypoint");
Object.defineProperty(exports, "EvaluatedEntrypoint", { enumerable: true, get: function () { return EvaluatedEntrypoint_1.EvaluatedEntrypoint; } });
var Entrypoint_helpers_1 = require("./transform/Entrypoint.helpers");
Object.defineProperty(exports, "parseFile", { enumerable: true, get: function () { return Entrypoint_helpers_1.parseFile; } });
var generators_1 = require("./transform/generators");
Object.defineProperty(exports, "baseHandlers", { enumerable: true, get: function () { return generators_1.baseHandlers; } });
var transform_2 = require("./transform/generators/transform");
Object.defineProperty(exports, "prepareCode", { enumerable: true, get: function () { return transform_2.prepareCode; } });
var Entrypoint_1 = require("./transform/Entrypoint");
Object.defineProperty(exports, "Entrypoint", { enumerable: true, get: function () { return Entrypoint_1.Entrypoint; } });
var createStylisPreprocessor_1 = require("./transform/generators/createStylisPreprocessor");
Object.defineProperty(exports, "transformUrl", { enumerable: true, get: function () { return createStylisPreprocessor_1.transformUrl; } });
var resolveImports_1 = require("./transform/generators/resolveImports");
Object.defineProperty(exports, "asyncResolveImports", { enumerable: true, get: function () { return resolveImports_1.asyncResolveImports; } });
Object.defineProperty(exports, "syncResolveImports", { enumerable: true, get: function () { return resolveImports_1.syncResolveImports; } });
var loadWywOptions_1 = require("./transform/helpers/loadWywOptions");
Object.defineProperty(exports, "loadWywOptions", { enumerable: true, get: function () { return loadWywOptions_1.loadWywOptions; } });
var withDefaultServices_1 = require("./transform/helpers/withDefaultServices");
Object.defineProperty(exports, "withDefaultServices", { enumerable: true, get: function () { return withDefaultServices_1.withDefaultServices; } });
var EventEmitter_1 = require("./utils/EventEmitter");
Object.defineProperty(exports, "EventEmitter", { enumerable: true, get: function () { return EventEmitter_1.EventEmitter; } });
var isNode_1 = require("./utils/isNode");
Object.defineProperty(exports, "isNode", { enumerable: true, get: function () { return isNode_1.isNode; } });
var getFileIdx_1 = require("./utils/getFileIdx");
Object.defineProperty(exports, "getFileIdx", { enumerable: true, get: function () { return getFileIdx_1.getFileIdx; } });
var getTagProcessor_1 = require("./utils/getTagProcessor");
Object.defineProperty(exports, "applyProcessors", { enumerable: true, get: function () { return getTagProcessor_1.applyProcessors; } });
var getVisitorKeys_1 = require("./utils/getVisitorKeys");
Object.defineProperty(exports, "getVisitorKeys", { enumerable: true, get: function () { return getVisitorKeys_1.getVisitorKeys; } });
var peek_1 = require("./utils/peek");
Object.defineProperty(exports, "peek", { enumerable: true, get: function () { return peek_1.peek; } });
var cache_1 = require("./cache");
Object.defineProperty(exports, "TransformCacheCollection", { enumerable: true, get: function () { return cache_1.TransformCacheCollection; } });
var findIdentifiers_1 = require("./utils/findIdentifiers");
Object.defineProperty(exports, "findIdentifiers", { enumerable: true, get: function () { return findIdentifiers_1.findIdentifiers; } });
