// @wyw-in-js/transform/esm/cache.js
import { createHash } from 'crypto';
import fs from 'node:fs';
import { logger } from '@wyw-in-js/shared';
import { getFileIdx } from './utils/getFileIdx';
import { stripQueryAndHash } from './utils/parseRequest';
function hashContent(content) {
  return createHash('sha256').update(content).digest('hex');
}
const cacheLogger = logger.extend('cache');
const cacheNames = ['entrypoints', 'exports'];
const loggers = cacheNames.reduce((acc, key) => ({
  ...acc,
  [key]: cacheLogger.extend(key)
}), {});
export class TransformCacheCollection {
  contentHashes = new Map();
  constructor(caches = {}) {
    this.entrypoints = caches.entrypoints || new Map();
    this.exports = caches.exports || new Map();
  }
  add(cacheName, key, value) {
    const cache = this[cacheName];
    loggers[cacheName]('%s:add %s %f', getFileIdx(key), key, () => {
      if (value === undefined) {
        return cache.has(key) ? 'removed' : 'noop';
      }
      if (!cache.has(key)) {
        return 'added';
      }
      return cache.get(key) === value ? 'unchanged' : 'updated';
    });
    if (value === undefined) {
      cache.delete(key);
      this.contentHashes.delete(key);
      return;
    }
    cache.set(key, value);
    if ('initialCode' in value) {
      this.contentHashes.set(key, hashContent(value.initialCode ?? ''));
    }
  }
  clear(cacheName) {
    if (cacheName === 'all') {
      cacheNames.forEach(name => {
        this.clear(name);
      });
      return;
    }
    loggers[cacheName]('clear');
    const cache = this[cacheName];
    cache.clear();
  }
  delete(cacheName, key) {
    this.invalidate(cacheName, key);
  }
  get(cacheName, key) {
    const cache = this[cacheName];
    const res = cache.get(key);
    loggers[cacheName]('get', key, res === undefined ? 'miss' : 'hit');
    return res;
  }
  has(cacheName, key) {
    const cache = this[cacheName];
    const res = cache.has(key);
    loggers[cacheName]('has', key, res);
    return res;
  }
  invalidate(cacheName, key) {
    const cache = this[cacheName];
    if (!cache.has(key)) {
      return;
    }
    loggers[cacheName]('invalidate', key);
    cache.delete(key);
  }
  invalidateForFile(filename) {
    cacheNames.forEach(cacheName => {
      this.invalidate(cacheName, filename);
    });
  }
  invalidateIfChanged(filename, content, previousVisitedFiles) {
    const visitedFiles = new Set(previousVisitedFiles);
    const fileEntrypoint = this.get('entrypoints', filename);

    // We need to check all dependencies of the file
    // because they might have changed as well.
    if (fileEntrypoint && !visitedFiles.has(filename)) {
      visitedFiles.add(filename);
      for (const [, dependency] of fileEntrypoint.dependencies) {
        const dependencyFilename = dependency.resolved;
        if (dependencyFilename) {
          const dependencyContent = fs.readFileSync(stripQueryAndHash(dependencyFilename), 'utf8');
          this.invalidateIfChanged(dependencyFilename, dependencyContent, visitedFiles);
        }
      }
    }
    const hash = this.contentHashes.get(filename);
    const newHash = hashContent(content);
    if (hash !== newHash) {
      cacheLogger('content has changed, invalidate all for %s', filename);
      this.contentHashes.set(filename, newHash);
      this.invalidateForFile(filename);
      return true;
    }
    return false;
  }
}
