// @wyw-in-js/transform/lib/transform/Entrypoint.helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMatchedRule = getMatchedRule;
exports.getStack = getStack;
exports.isSuperSet = void 0;
exports.loadAndParse = loadAndParse;
exports.mergeOnly = mergeOnly;
exports.parseFile = parseFile;
var _fs = require("fs");
var _path = require("path");
var _shared = require("@wyw-in-js/shared");
var _buildOptions = require("../options/buildOptions");
var _loadBabelOptions = require("../options/loadBabelOptions");
var _getFileIdx = require("../utils/getFileIdx");
var _getPluginKey = require("../utils/getPluginKey");
var _parseRequest = require("../utils/parseRequest");
function getMatchedRule(rules, filename, code) {
  for (let i = rules.length - 1; i >= 0; i--) {
    const rule = rules[i];
    if (!rule.test) {
      return rule;
    }
    if (typeof rule.test === 'function' && rule.test(filename, code)) {
      return rule;
    }
    if (rule.test instanceof RegExp && rule.test.test(filename)) {
      return rule;
    }
  }
  return {
    action: 'ignore'
  };
}
function parseFile(babel, filename, originalCode, parseConfig) {
  const log = _shared.logger.extend('transform:parse').extend((0, _getFileIdx.getFileIdx)(filename));
  const parseResult = babel.parseSync(originalCode, parseConfig);
  if (!parseResult) {
    throw new Error(`Failed to parse ${filename}`);
  }
  log('stage-1', `${filename} has been parsed`);
  return parseResult;
}
const isModuleResolver = plugin => {
  const key = (0, _getPluginKey.getPluginKey)(plugin);
  if (!key) return false;
  if (['module-resolver', 'babel-plugin-module-resolver'].includes(key)) {
    return true;
  }
  return /([\\/])babel-plugin-module-resolver\1/.test(key);
};
let moduleResolverWarned = false;
const normalizeBabelKey = key => key.replace(/\\/g, '/');
const isBabelPresetTypescript = key => {
  const normalized = normalizeBabelKey(key);
  if (normalized === 'typescript') return true;
  return normalized.includes('preset-typescript');
};
const isBabelTransformTypescriptPlugin = key => {
  const normalized = normalizeBabelKey(key);
  if (normalized === 'transform-typescript') return true;
  return normalized.includes('plugin-transform-typescript');
};
const withAllowDeclareFields = item => {
  if (!Array.isArray(item)) {
    return [item, {
      allowDeclareFields: true
    }];
  }
  const [target, rawOptions, ...rest] = item;
  const options = typeof rawOptions === 'object' && rawOptions !== null && !Array.isArray(rawOptions) ? rawOptions : {};
  if ('allowDeclareFields' in options) {
    return item;
  }
  return [target, {
    ...options,
    allowDeclareFields: true
  }, ...rest];
};
const ensureAllowDeclareFieldsInBabelOptions = (babelOptions, scope = 'top') => {
  var _babelOptions$presets, _babelOptions$plugins;
  let presetsChanged = false;
  let pluginsChanged = false;
  let overridesChanged = false;
  let envChanged = false;
  const presets = (_babelOptions$presets = babelOptions.presets) === null || _babelOptions$presets === void 0 ? void 0 : _babelOptions$presets.map(item => {
    const key = (0, _getPluginKey.getPluginKey)(item);
    if (!key || !isBabelPresetTypescript(key)) {
      return item;
    }
    presetsChanged = true;
    return withAllowDeclareFields(item);
  });
  const plugins = (_babelOptions$plugins = babelOptions.plugins) === null || _babelOptions$plugins === void 0 ? void 0 : _babelOptions$plugins.map(item => {
    const key = (0, _getPluginKey.getPluginKey)(item);
    if (!key || !isBabelTransformTypescriptPlugin(key)) {
      return item;
    }
    pluginsChanged = true;
    return withAllowDeclareFields(item);
  });
  const {
    overrides: baseOverrides
  } = babelOptions;
  let overrides = baseOverrides;
  if (scope === 'top' && baseOverrides) {
    const patchedOverrides = baseOverrides.map(override => ensureAllowDeclareFieldsInBabelOptions(override, 'override'));
    if (patchedOverrides.some((patchedOverride, idx) => patchedOverride !== baseOverrides[idx])) {
      overridesChanged = true;
      overrides = patchedOverrides;
    }
  }
  const {
    env: baseEnv
  } = babelOptions;
  let env = baseEnv;
  if (scope === 'top' && baseEnv) {
    const entries = Object.entries(baseEnv);
    const patchedEntries = entries.map(([envName, envOptions]) => [envName, envOptions ? ensureAllowDeclareFieldsInBabelOptions(envOptions, 'env') : envOptions]);
    if (patchedEntries.some(([, patched], idx) => patched !== entries[idx][1])) {
      envChanged = true;
      env = Object.fromEntries(patchedEntries);
    }
  }
  if (!presetsChanged && !pluginsChanged && !overridesChanged && !envChanged) {
    return babelOptions;
  }
  const next = {
    ...babelOptions
  };
  if (presetsChanged) next.presets = presets;
  if (pluginsChanged) next.plugins = plugins;
  if (overridesChanged) next.overrides = overrides;
  if (envChanged) next.env = env;
  return next;
};
function buildConfigs(services, name, pluginOptions, babelOptions) {
  var _parseConfig$plugins, _rawConfig$plugins;
  const {
    babel,
    options
  } = services;
  const commonOptions = {
    ast: true,
    filename: name,
    inputSourceMap: options.inputSourceMap,
    root: options.root,
    sourceFileName: name,
    sourceMaps: true
  };
  const isTypescriptFile = name.endsWith('.ts') || name.endsWith('.tsx') || name.endsWith('.mts') || name.endsWith('.cts');
  let rawConfig = (0, _buildOptions.buildOptions)(pluginOptions === null || pluginOptions === void 0 ? void 0 : pluginOptions.babelOptions, babelOptions, commonOptions);
  if (isTypescriptFile) {
    rawConfig = ensureAllowDeclareFieldsInBabelOptions(rawConfig);
  }
  const useBabelConfigs = (0, _shared.isFeatureEnabled)(pluginOptions.features, 'useBabelConfigs', name);
  if (!useBabelConfigs) {
    rawConfig = {
      ...rawConfig,
      configFile: false
    };
  }
  const parseConfig = (0, _loadBabelOptions.loadBabelOptions)(babel, name, {
    babelrc: useBabelConfigs,
    ...rawConfig
  });
  const parseHasModuleResolver = (_parseConfig$plugins = parseConfig.plugins) === null || _parseConfig$plugins === void 0 ? void 0 : _parseConfig$plugins.some(isModuleResolver);
  const rawHasModuleResolver = (_rawConfig$plugins = rawConfig.plugins) === null || _rawConfig$plugins === void 0 ? void 0 : _rawConfig$plugins.some(isModuleResolver);
  if (parseHasModuleResolver && !rawHasModuleResolver) {
    var _parseConfig$plugins$, _parseConfig$plugins2, _rawConfig$plugins2;
    if (!moduleResolverWarned) {
      // eslint-disable-next-line no-console
      console.warn(`[wyw-in-js] ${name} has a module-resolver plugin in its babelrc, but it is not present ` + `in the babelOptions for the wyw-in-js plugin. This works for now but will be an error in the future. ` + `Please add the module-resolver plugin to the babelOptions for the wyw-in-js plugin.`);
      moduleResolverWarned = true;
    }
    rawConfig = {
      ...rawConfig,
      plugins: [...((_parseConfig$plugins$ = (_parseConfig$plugins2 = parseConfig.plugins) === null || _parseConfig$plugins2 === void 0 ? void 0 : _parseConfig$plugins2.filter(plugin => isModuleResolver(plugin))) !== null && _parseConfig$plugins$ !== void 0 ? _parseConfig$plugins$ : []), ...((_rawConfig$plugins2 = rawConfig.plugins) !== null && _rawConfig$plugins2 !== void 0 ? _rawConfig$plugins2 : [])]
    };
  }
  const evalConfig = (0, _loadBabelOptions.loadBabelOptions)(babel, name, {
    babelrc: false,
    ...rawConfig
  });
  return {
    evalConfig,
    parseConfig
  };
}
function loadAndParse(services, name, loadedCode, log) {
  const {
    babel,
    eventEmitter,
    options: {
      pluginOptions
    }
  } = services;
  const filename = (0, _parseRequest.stripQueryAndHash)(name);
  const extension = (0, _path.extname)(filename);
  if (!pluginOptions.extensions.includes(extension)) {
    log('[createEntrypoint] %s is ignored. If you want it to be processed, you should add \'%s\' to the "extensions" option.', filename, extension);
    return {
      code: (0, _path.isAbsolute)(filename) ? loadedCode : '',
      evaluator: 'ignored',
      reason: 'extension'
    };
  }
  const code = loadedCode !== null && loadedCode !== void 0 ? loadedCode : (0, _fs.readFileSync)(filename, 'utf-8');
  const {
    action,
    babelOptions
  } = getMatchedRule(pluginOptions.rules, filename, code);
  let ast;
  const {
    evalConfig,
    parseConfig
  } = buildConfigs(services, filename, pluginOptions, babelOptions);
  const getOrParse = () => {
    if (ast) return ast;
    ast = eventEmitter.perf('parseFile', () => parseFile(babel, name, code, parseConfig));
    return ast;
  };
  if (action === 'ignore') {
    log('[createEntrypoint] %s is ignored by rule', name);
    return {
      get ast() {
        return getOrParse();
      },
      code,
      evaluator: 'ignored',
      reason: 'rule'
    };
  }
  const evaluator = typeof action === 'function' ? action : require(require.resolve(action, {
    paths: [(0, _path.dirname)(filename)]
  })).default;
  return {
    get ast() {
      return getOrParse();
    },
    code,
    evaluator,
    evalConfig
  };
}
function getStack(entrypoint) {
  if (!entrypoint) return [];
  const stack = [entrypoint.name];
  let {
    parents
  } = entrypoint;
  while (parents.length) {
    stack.push(parents[0].name);
    parents = parents[0].parents;
  }
  return stack;
}
function mergeOnly(a, b) {
  const result = new Set(a);
  b.forEach(item => result.add(item));
  return [...result].filter(i => i).sort();
}
const isSuperSet = (a, b) => {
  if (a.includes('*')) return true;
  if (b.length === 0) return true;
  const aSet = new Set(a);
  return b.every(item => aSet.has(item));
};
exports.isSuperSet = isSuperSet;
