// @wyw-in-js/transform/lib/transform/isStaticallyEvaluatableModule.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isStaticallyEvaluatableModule = isStaticallyEvaluatableModule;
var _types = require("@babel/types");
function isTypeOnlyImport(statement) {
  if (statement.importKind === 'type') {
    return true;
  }
  return statement.specifiers.every(specifier => (0, _types.isImportSpecifier)(specifier) && specifier.importKind === 'type');
}
function isTypeOnlyReExport(statement) {
  if (!statement.source) {
    return false;
  }
  return statement.exportKind === 'type';
}
function unwrapExpression(expr) {
  let current = expr;
  for (;;) {
    if ((0, _types.isTSAsExpression)(current)) {
      current = current.expression;
      // eslint-disable-next-line no-continue
      continue;
    }
    if ((0, _types.isTSSatisfiesExpression)(current)) {
      current = current.expression;
      // eslint-disable-next-line no-continue
      continue;
    }
    if ((0, _types.isTSNonNullExpression)(current)) {
      current = current.expression;
      // eslint-disable-next-line no-continue
      continue;
    }
    if ((0, _types.isTSInstantiationExpression)(current)) {
      current = current.expression;
      // eslint-disable-next-line no-continue
      continue;
    }
    if ((0, _types.isParenthesizedExpression)(current)) {
      current = current.expression;
      // eslint-disable-next-line no-continue
      continue;
    }
    return current;
  }
}
function isSafeExpression(expr) {
  const unwrapped = unwrapExpression(expr);
  if ((0, _types.isStringLiteral)(unwrapped) || (0, _types.isNumericLiteral)(unwrapped) || (0, _types.isBooleanLiteral)(unwrapped) || (0, _types.isNullLiteral)(unwrapped) || (0, _types.isBigIntLiteral)(unwrapped)) {
    return true;
  }
  if ((0, _types.isArrowFunctionExpression)(unwrapped) || (0, _types.isFunctionExpression)(unwrapped)) {
    return true;
  }
  if ((0, _types.isIdentifier)(unwrapped)) {
    return unwrapped.name === 'undefined' || unwrapped.name === 'NaN' || unwrapped.name === 'Infinity';
  }
  if ((0, _types.isTemplateLiteral)(unwrapped)) {
    return unwrapped.expressions.every(item => (0, _types.isExpression)(item) && isSafeExpression(item));
  }
  if ((0, _types.isUnaryExpression)(unwrapped)) {
    return isSafeExpression(unwrapped.argument);
  }
  if ((0, _types.isBinaryExpression)(unwrapped) || (0, _types.isLogicalExpression)(unwrapped)) {
    return isSafeExpression(unwrapped.left) && isSafeExpression(unwrapped.right);
  }
  if ((0, _types.isConditionalExpression)(unwrapped)) {
    return isSafeExpression(unwrapped.test) && isSafeExpression(unwrapped.consequent) && isSafeExpression(unwrapped.alternate);
  }
  if ((0, _types.isArrayExpression)(unwrapped)) {
    return unwrapped.elements.every(item => {
      if (item === null) return true;
      if ((0, _types.isSpreadElement)(item)) return false;
      return isSafeExpression(item);
    });
  }
  if ((0, _types.isObjectExpression)(unwrapped)) {
    return unwrapped.properties.every(prop => {
      if ((0, _types.isSpreadElement)(prop)) {
        return false;
      }
      if ((0, _types.isObjectMethod)(prop)) {
        return !prop.computed;
      }
      if ((0, _types.isObjectProperty)(prop)) {
        if (prop.computed) {
          return false;
        }
        return (0, _types.isExpression)(prop.value) && isSafeExpression(prop.value);
      }
      return false;
    });
  }
  if ((0, _types.isCallExpression)(unwrapped) || (0, _types.isNewExpression)(unwrapped) || (0, _types.isTaggedTemplateExpression)(unwrapped) || (0, _types.isAwaitExpression)(unwrapped) || (0, _types.isYieldExpression)(unwrapped) || (0, _types.isUpdateExpression)(unwrapped) || (0, _types.isAssignmentExpression)(unwrapped) || (0, _types.isSequenceExpression)(unwrapped)) {
    return false;
  }
  return false;
}
function isSafeDeclarator(declarator) {
  if (!declarator.init) {
    return true;
  }
  return isSafeExpression(declarator.init);
}
function isSafeVariableDeclaration(decl) {
  return decl.declarations.every(isSafeDeclarator);
}
function isSafeStatement(statement) {
  if ((0, _types.isImportDeclaration)(statement)) {
    return isTypeOnlyImport(statement);
  }
  if ((0, _types.isExportAllDeclaration)(statement)) {
    return false;
  }
  if ((0, _types.isExportNamedDeclaration)(statement)) {
    if (!statement.declaration) {
      return !statement.source || isTypeOnlyReExport(statement);
    }
    if ((0, _types.isFunctionDeclaration)(statement.declaration)) {
      return true;
    }
    if ((0, _types.isClassDeclaration)(statement.declaration)) {
      return false;
    }
    if (statement.declaration.type === 'VariableDeclaration') {
      return isSafeVariableDeclaration(statement.declaration);
    }
    return false;
  }
  if ((0, _types.isExportDefaultDeclaration)(statement)) {
    const decl = statement.declaration;
    if ((0, _types.isFunctionDeclaration)(decl) || (0, _types.isFunctionExpression)(decl) || (0, _types.isArrowFunctionExpression)(decl) || (0, _types.isClassExpression)(decl) || (0, _types.isClassDeclaration)(decl)) {
      return false;
    }
    return isSafeExpression(decl);
  }
  if (statement.type === 'VariableDeclaration') {
    return isSafeVariableDeclaration(statement);
  }
  if ((0, _types.isFunctionDeclaration)(statement)) {
    return true;
  }
  if ((0, _types.isEmptyStatement)(statement)) {
    return true;
  }
  if ((0, _types.isExpressionStatement)(statement)) {
    // Directives (like "use strict") are safe.
    return (0, _types.isStringLiteral)(statement.expression);
  }
  return false;
}
function isStaticallyEvaluatableModule(ast) {
  return ast.program.body.every(isSafeStatement);
}
