// @wyw-in-js/transform/lib/plugins/shaker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = shakerPlugin;
var _path = _interopRequireDefault(require("path"));
var _shared = require("@wyw-in-js/shared");
var _collectExportsAndImports = require("../utils/collectExportsAndImports");
var _getFileIdx = require("../utils/getFileIdx");
var _isRemoved = require("../utils/isRemoved");
var _scopeHelpers = require("../utils/scopeHelpers");
var _traversalCache = require("../utils/traversalCache");
var _parseRequest = require("../utils/parseRequest");
var _importOverrides = require("../utils/importOverrides");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function getNonParamBinding(exportPath, name) {
  var _exportPath$scope$par, _exportPath$scope$par2;
  const binding = exportPath.scope.getBinding(name);
  if (binding && binding.kind !== 'param') {
    return binding;
  }

  // When `exportPath` is inside a function scope, a parameter can shadow
  // the actual export binding (e.g. `export function fallback(fallback) {}`).
  // In such cases we need the binding from the declaration scope.
  return (_exportPath$scope$par = (_exportPath$scope$par2 = exportPath.scope.parent) === null || _exportPath$scope$par2 === void 0 ? void 0 : _exportPath$scope$par2.getBinding(name)) !== null && _exportPath$scope$par !== void 0 ? _exportPath$scope$par : binding;
}
function getBindingForExport(exportPath) {
  if (exportPath.isIdentifier()) {
    return getNonParamBinding(exportPath, exportPath.node.name);
  }
  const variableDeclarator = exportPath.findParent(p => p.isVariableDeclarator());
  if (variableDeclarator) {
    const id = variableDeclarator.get('id');
    if (id.isIdentifier()) {
      return variableDeclarator.scope.getBinding(id.node.name);
    }
  }
  if (exportPath.isAssignmentExpression()) {
    const left = exportPath.get('left');
    if (left.isIdentifier()) {
      return getNonParamBinding(exportPath, left.node.name);
    }
  }
  if (exportPath.isFunctionDeclaration() || exportPath.isClassDeclaration()) {
    const {
      id
    } = exportPath.node;
    if (!id) {
      // `export default function() {}` / `export default class {}` (anonymous)
      return undefined;
    }
    return getNonParamBinding(exportPath, id.name);
  }
  return undefined;
}
const withoutRemoved = items => items.filter(({
  local
}) => !(0, _isRemoved.isRemoved)(local));
function rearrangeExports({
  types: t
}, root, exportRefs, exports) {
  const rearranged = {
    ...exports
  };
  const rootScope = root.scope;
  exportRefs.forEach((refs, name) => {
    var _constantViolations;
    if (refs.length <= 1) {
      if (refs.length === 1) {
        // Maybe exports is assigned to another variable?
        const declarator = refs[0].findParent(p => p.isVariableDeclarator());
        if (!declarator) {
          return;
        }
      } else {
        return;
      }
    }
    const uid = rootScope.generateUid(name);
    // Define variable in the beginning
    const [declaration] = root.unshiftContainer('body', [t.variableDeclaration('var', [t.variableDeclarator(t.identifier(uid))])]);
    rootScope.registerDeclaration(declaration);
    const constantViolations = [];
    // Replace every reference with defined variable
    refs.forEach(ref => {
      const [replaced] = ref.replaceWith(t.identifier(uid));
      if (replaced.isBindingIdentifier()) {
        constantViolations.push(replaced);
      } else {
        (0, _scopeHelpers.reference)(replaced);
      }
    });
    constantViolations.forEach(id => {
      rootScope.registerConstantViolation(id);
    });
    const assigmentToExport = t.expressionStatement(t.assignmentExpression('=', t.memberExpression(t.identifier('exports'), t.identifier(name)), t.identifier(uid)));

    // export.foo = _foo will be inserted either after the last _foo assigment or in the end of the file
    const body = root.get('body');
    const lastViolation = (_constantViolations = constantViolations[constantViolations.length - 1]) !== null && _constantViolations !== void 0 ? _constantViolations : body[body.length - 1];
    const pathInRoot = root.get('body').find(n => lastViolation.isDescendant(n));
    const [pushed] = pathInRoot.insertAfter(assigmentToExport);
    const local = pushed.get('expression.right');
    (0, _scopeHelpers.reference)(local);
    rearranged[name] = local;
  });
  return rearranged;
}
const getPropertyAssignmentStatement = (ref, bindingName) => {
  const assignment = ref.findParent(parent => parent.isAssignmentExpression());
  if (!(assignment !== null && assignment !== void 0 && assignment.isAssignmentExpression())) return null;
  const left = assignment.get('left');
  if (!left.isMemberExpression()) return null;
  const object = left.get('object');
  if (!object.isIdentifier({
    name: bindingName
  })) return null;
  const statement = assignment.parentPath;
  return statement !== null && statement !== void 0 && statement.isExpressionStatement() ? statement : null;
};
const isWithinAliveExport = (ref, aliveExports) => [...aliveExports].some(alive => alive === ref || alive.isAncestor(ref));
function stripExportKeepDeclaration(path) {
  const exportDeclaration = path.findParent(p => p.isExportNamedDeclaration());
  if (!exportDeclaration) return false;
  const declaration = exportDeclaration.get('declaration');
  if (!declaration.node) return false;
  if (declaration.isFunctionDeclaration() || declaration.isClassDeclaration() || declaration.isTSEnumDeclaration()) {
    exportDeclaration.replaceWith(declaration.node);
    return true;
  }
  if (declaration.isVariableDeclaration()) {
    const declarators = declaration.get('declarations');
    if (declarators.length !== 1) {
      return false;
    }
    exportDeclaration.replaceWith(declaration.node);
    return true;
  }
  return false;
}
function shakerPlugin(babel, {
  keepSideEffects = false,
  ifUnknownExport = 'skip-shaking',
  importOverrides,
  onlyExports,
  root
}) {
  const shakerLogger = _shared.logger.extend('shaker');
  return {
    name: '@wyw-in-js/transform/shaker',
    pre(file) {
      this.filename = file.opts.filename;
      const log = shakerLogger.extend((0, _getFileIdx.getFileIdx)(this.filename));
      log('start', `${this.filename}, onlyExports: ${onlyExports.join(',')}`);
      const onlyExportsSet = new Set(onlyExports);
      const shouldKeepOverride = override => !!override && ('mock' in override || 'noShake' in override);
      const hasImportOverride = (() => {
        if (!importOverrides || Object.keys(importOverrides).length === 0) {
          return () => false;
        }
        const cache = new Map();
        return source => {
          var _importOverrides$sour;
          const cached = cache.get(source);
          if (cached !== undefined) {
            return cached;
          }
          const strippedSource = (0, _parseRequest.stripQueryAndHash)(source);
          const direct = (_importOverrides$sour = importOverrides[source]) !== null && _importOverrides$sour !== void 0 ? _importOverrides$sour : strippedSource !== source ? importOverrides[strippedSource] : null;
          if (direct) {
            const result = shouldKeepOverride(direct);
            cache.set(source, result);
            return result;
          }
          const isFileImport = strippedSource.startsWith('.') || _path.default.isAbsolute(strippedSource);
          if (!isFileImport) {
            cache.set(source, false);
            return false;
          }
          try {
            const resolved = (0, _shared.syncResolve)(strippedSource, this.filename, []);
            const {
              key
            } = (0, _importOverrides.toImportKey)({
              source: strippedSource,
              resolved,
              root
            });
            const override = importOverrides[key];
            const result = shouldKeepOverride(override);
            cache.set(source, result);
            return result;
          } catch {
            cache.set(source, false);
            return false;
          }
        };
      })();
      const collected = (0, _collectExportsAndImports.collectExportsAndImports)(file.path);
      const sideEffectImports = collected.imports.filter(_collectExportsAndImports.sideEffectImport);
      log('import-and-exports', [`imports: ${collected.imports.length} (side-effects: ${sideEffectImports.length})`, `exports: ${Object.values(collected.exports).length}`, `reexports: ${collected.reexports.length}`].join(', '));

      // We cannot just throw out exports if they are referred in the code
      // Let's dome some replacements
      const exports = rearrangeExports(babel, file.path, collected.exportRefs, collected.exports);
      Object.values(exports).forEach(local => {
        if (local.isAssignmentExpression()) {
          const left = local.get('left');
          if (left.isIdentifier()) {
            // For some reason babel does not mark id in AssignmentExpression as a reference
            // So we need to do it manually
            (0, _scopeHelpers.reference)(left, left, true);
          }
        }
      });
      const hasWywPreval = exports.__wywPreval !== undefined;
      const hasDefault = exports.default !== undefined;

      // If __wywPreval is not exported, we can remove it from onlyExports
      if (onlyExportsSet.has('__wywPreval') && !hasWywPreval) {
        onlyExportsSet.delete('__wywPreval');
      }
      if (onlyExportsSet.size === 0) {
        // Fast-lane: if there are no exports to keep, we can just shake out the whole file
        this.imports = [];
        this.exports = {};
        this.reexports = [];
        this.deadExports = Object.keys(exports);
        file.path.get('body').forEach(p => {
          p.remove();
        });
        return;
      }
      const importedAsSideEffect = onlyExportsSet.has('side-effect');
      onlyExportsSet.delete('side-effect');

      // Hackaround for packages which include a 'default' export without specifying __esModule; such packages cannot be
      // shaken as they will break interopRequireDefault babel helper
      // See example in shaker-plugin.test.ts
      // Real-world example was found in preact/compat npm package
      if (onlyExportsSet.has('default') && hasDefault && !collected.isEsModule) {
        this.imports = collected.imports;
        this.exports = exports;
        this.reexports = collected.reexports;
        this.deadExports = [];
        return;
      }
      if (!onlyExportsSet.has('*')) {
        // __esModule should be kept alive
        onlyExportsSet.add('__esModule');
        const aliveExports = new Set();
        const importNames = collected.imports.map(({
          imported
        }) => imported);
        Object.entries(exports).forEach(([exported, local]) => {
          if (onlyExportsSet.has(exported)) {
            aliveExports.add(local);
            return;
          }
          const binding = local.isIdentifier() && local.scope.getBinding(local.node.name);
          if (binding && (binding.path.isImportSpecifier() || binding.path.isImportDefaultSpecifier() || binding.path.isImportNamespaceSpecifier()) && importNames.includes(local.node.name || '')) {
            aliveExports.add(local);
            return;
          }
          if ([...aliveExports].some(alive => alive === local)) {
            // It's possible to export multiple values from a single variable initializer, e.g
            // export const { foo, bar } = baz();
            // We need to treat all of them as used if any of them are used, since otherwise
            // we'll attempt to delete the baz() call
            aliveExports.add(local);
          }
        });
        collected.reexports.forEach(exp => {
          if (onlyExportsSet.has(exp.exported)) {
            aliveExports.add(exp.local);
          }
        });
        const exportToPath = new Map();
        Object.entries(exports).forEach(([exported, local]) => {
          exportToPath.set(exported, local);
        });
        collected.reexports.forEach(exp => {
          exportToPath.set(exp.exported, exp.local);
        });
        const notFoundExports = [...onlyExportsSet].filter(exp => exp !== '__esModule' && !aliveExports.has(exportToPath.get(exp)));
        exportToPath.clear();
        const isAllExportsFound = notFoundExports.length === 0;
        if (!isAllExportsFound && ifUnknownExport !== 'ignore') {
          if (ifUnknownExport === 'error') {
            throw new Error(`Unknown export(s) requested: ${onlyExports.join(',')}`);
          }
          if (ifUnknownExport === 'reexport-all') {
            // If there are unknown exports, we have keep alive all re-exports.
            if (exports['*'] !== undefined) {
              aliveExports.add(exports['*']);
            }
            collected.reexports.forEach(exp => {
              if (exp.exported === '*') {
                aliveExports.add(exp.local);
              }
            });
          }
          if (ifUnknownExport === 'skip-shaking') {
            this.imports = collected.imports;
            this.exports = exports;
            this.reexports = collected.reexports;
            this.deadExports = [];
            return;
          }
        }
        const forDeleting = [...Object.values(exports), ...collected.reexports.map(i => i.local)].filter(exp => !aliveExports.has(exp));
        const forDeletingSet = new Set(forDeleting);
        const queueForDeleting = path => {
          if ((0, _isRemoved.isRemoved)(path) || forDeletingSet.has(path)) {
            return false;
          }
          forDeletingSet.add(path);
          forDeleting.push(path);
          return true;
        };
        if (!keepSideEffects && !importedAsSideEffect) {
          // Drop side-effect imports for eval-only builds unless they were explicitly requested.
          // This prevents evaluating unrelated runtime code (e.g. Radix) during __wywPreval eval.
          sideEffectImports.forEach(i => {
            if (hasImportOverride(i.source)) {
              return;
            }
            queueForDeleting(i.local);
          });
        }
        const deleted = new Set();
        let dereferenced = [];
        let changed = true;
        while (changed) {
          changed = false;
          // eslint-disable-next-line no-restricted-syntax
          for (const path of forDeleting) {
            if (deleted.has(path)) {
              // eslint-disable-next-line no-continue
              continue;
            }
            const binding = getBindingForExport(path);
            const action = (0, _scopeHelpers.findActionForNode)(path);
            const parent = action === null || action === void 0 ? void 0 : action[1];
            const outerReferences = ((binding === null || binding === void 0 ? void 0 : binding.referencePaths) || []).filter(ref => {
              if (ref === parent || parent !== null && parent !== void 0 && parent.isAncestor(ref)) {
                return false;
              }
              return !forDeleting.some(candidate => candidate !== path && !(0, _isRemoved.isRemoved)(candidate) && (candidate === ref || candidate.isAncestor(ref)));
            });
            const bindingName = binding === null || binding === void 0 ? void 0 : binding.identifier.name;
            const removableAssignmentStatements = new Set();
            const removableOuterReferences = outerReferences.filter(ref => {
              if (!bindingName) return false;
              const statement = getPropertyAssignmentStatement(ref, bindingName);
              if (!statement || isWithinAliveExport(statement, aliveExports)) {
                return false;
              }
              removableAssignmentStatements.add(statement);
              return true;
            });
            const blockingReferences = outerReferences.filter(ref => !removableOuterReferences.includes(ref));
            if (blockingReferences.length > 0 && path.isIdentifier()) {
              // Temporary deref it in order to simplify further checks.
              (0, _scopeHelpers.dereference)(path);
              dereferenced.push(path);
            }
            if (!deleted.has(path) && binding && blockingReferences.length > 0 && stripExportKeepDeclaration(path)) {
              deleted.add(path);
              changed = true;
              // eslint-disable-next-line no-continue
              continue;
            }
            if (!deleted.has(path) && (!binding || blockingReferences.length === 0)) {
              if (removableAssignmentStatements.size > 0) {
                for (const statement of removableAssignmentStatements) {
                  if (queueForDeleting(statement)) {
                    changed = true;
                  }
                }
              }
              if (action) {
                (0, _scopeHelpers.applyAction)(action);
              } else {
                (0, _scopeHelpers.removeWithRelated)([path]);
              }
              deleted.add(path);
              changed = true;
            }
          }
          dereferenced.forEach(path => {
            // If path is still alive, we need to reference it back
            if (!(0, _isRemoved.isRemoved)(path)) {
              (0, _scopeHelpers.reference)(path);
            }
          });
          dereferenced = [];

          // Find and mark for deleting all unreferenced variables
          const unreferenced = Object.values(file.scope.getAllBindings()).filter(i => !i.referenced);
          for (const binding of unreferenced) {
            if (binding.path.isVariableDeclarator()) {
              const id = binding.path.get('id');
              if (!(0, _isRemoved.isRemoved)(id) && !forDeletingSet.has(id)) {
                // Drop dead variable declarations, e.g. `const foo = make();` when `foo` is no longer referenced.
                for (const violation of binding.constantViolations) {
                  if (queueForDeleting(violation)) {
                    changed = true;
                  }
                }
                if (queueForDeleting(id)) {
                  changed = true;
                }
              }
            }

            // Drop import specifiers whose bindings lost all references during shaking
            // (e.g. when we keep only __wywPreval and the rest of the module is removed).
            if ((binding.path.isImportSpecifier() || binding.path.isImportDefaultSpecifier() || binding.path.isImportNamespaceSpecifier()) && !(0, _isRemoved.isRemoved)(binding.path) && !forDeletingSet.has(binding.path)) {
              if (queueForDeleting(binding.path)) {
                changed = true;
              }
            }
          }
        }
      }
      this.imports = withoutRemoved(collected.imports);
      this.exports = {};
      this.deadExports = [];
      Object.entries(exports).forEach(([exported, local]) => {
        if ((0, _isRemoved.isRemoved)(local)) {
          this.deadExports.push(exported);
        } else {
          this.exports[exported] = local;
        }
      });
      this.reexports = withoutRemoved(collected.reexports);
    },
    visitor: {},
    post(file) {
      const log = shakerLogger.extend((0, _getFileIdx.getFileIdx)(file.opts.filename));
      const processedImports = new Set();
      const imports = new Map();
      const addImport = ({
        imported,
        source
      }) => {
        if (processedImports.has(`${source}:${imported}`)) {
          return;
        }
        if (!imports.has(source)) {
          imports.set(source, []);
        }
        if (imported) {
          imports.get(source).push(imported);
        }
        processedImports.add(`${source}:${imported}`);
      };
      this.imports.forEach(addImport);
      this.reexports.forEach(addImport);
      log('end', `remaining imports: %O`, imports);

      // eslint-disable-next-line no-param-reassign
      file.metadata.wywEvaluator = {
        imports
      };
      (0, _traversalCache.invalidateTraversalCache)(file.path);
    }
  };
}
