// @wyw-in-js/vite/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = wywInJS;
var _fs = require("fs");
var _path = _interopRequireDefault(require("path"));
var _vite = require("vite");
var _shared = require("@wyw-in-js/shared");
var _transform = require("@wyw-in-js/transform");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * This file contains a Vite loader for wyw-in-js.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */

const isWindowsAbsolutePath = value => /^[a-zA-Z]:[\\/]/.test(value);
const normalizeToPosix = value => value.replace(/\\/g, _path.default.posix.sep);
const isInside = (childPath, parentPath) => {
  const rel = _path.default.relative(parentPath, childPath);
  return rel === '' || !rel.startsWith('..') && !_path.default.isAbsolute(rel);
};
const isWywCssAssetName = value => value.endsWith('.wyw-in-js.css');
const normalizeAssetRelativePath = value => {
  const normalized = _path.default.posix.normalize(normalizeToPosix(value).replace(/^\/+/, ''));
  if (normalized.startsWith('..') || _path.default.posix.isAbsolute(normalized)) {
    return null;
  }
  return normalized;
};
const stripExtension = value => {
  const ext = _path.default.posix.extname(value);
  return ext ? value.slice(0, -ext.length) : value;
};
const VITE_FS_PREFIX = '/@fs/';
const safeDecodeURIComponent = value => {
  try {
    return decodeURIComponent(value);
  } catch {
    return value;
  }
};
const normalizeViteFsPath = value => {
  const fsPath = value.slice(VITE_FS_PREFIX.length);
  return _path.default.normalize(safeDecodeURIComponent(fsPath));
};
const getWywCssAssetFileNames = (resolvedConfig, output, originalAssetFileNames) => {
  if (!output.preserveModules) return null;
  const rootDir = resolvedConfig.root;
  const preserveModulesRootValue = output.preserveModulesRoot;
  let preserveModulesRootAbs = null;
  if (typeof preserveModulesRootValue === 'string') {
    preserveModulesRootAbs = _path.default.isAbsolute(preserveModulesRootValue) ? preserveModulesRootValue : _path.default.resolve(rootDir, preserveModulesRootValue);
  }
  const preserveModulesRootRel = preserveModulesRootAbs && isInside(preserveModulesRootAbs, rootDir) ? normalizeToPosix(_path.default.relative(rootDir, preserveModulesRootAbs)) : null;
  return assetInfo => {
    const template = typeof originalAssetFileNames === 'function' ? originalAssetFileNames(assetInfo) : originalAssetFileNames;
    const assetName = assetInfo === null || assetInfo === void 0 ? void 0 : assetInfo.name;
    if (typeof assetName !== 'string' || !isWywCssAssetName(assetName)) {
      return template;
    }
    if (!template.includes('[')) {
      return template;
    }
    let relativePath = null;
    const assetNameNormalized = normalizeToPosix(assetName);
    if (_path.default.isAbsolute(assetName) || isWindowsAbsolutePath(assetNameNormalized)) {
      const preserveRel = preserveModulesRootAbs && isInside(assetName, preserveModulesRootAbs) ? _path.default.relative(preserveModulesRootAbs, assetName) : null;
      if (preserveRel && !_path.default.isAbsolute(preserveRel) && !preserveRel.startsWith('..')) {
        relativePath = preserveRel;
      } else if (isInside(assetName, rootDir)) {
        relativePath = _path.default.relative(rootDir, assetName);
      }
    } else if (preserveModulesRootRel && assetNameNormalized.startsWith(`${preserveModulesRootRel}/`)) {
      relativePath = assetNameNormalized.slice(preserveModulesRootRel.length + 1);
    } else {
      relativePath = assetNameNormalized;
    }
    const normalized = relativePath ? normalizeAssetRelativePath(relativePath) : null;
    if (!normalized) {
      return template;
    }
    const withoutExt = stripExtension(normalized);
    if (template.includes('[name]')) {
      const dir = _path.default.posix.dirname(withoutExt);
      if (dir === '.' || dir === '') {
        return template;
      }
      return template.replace(/\[name\]/g, `${dir}/[name]`);
    }
    const dir = _path.default.posix.dirname(withoutExt);
    if (dir === '.' || dir === '') {
      return template;
    }
    const idx = template.indexOf('[');
    if (idx < 0) {
      return template;
    }
    const prefix = template.slice(0, idx);
    if (prefix !== '' && !prefix.endsWith('/')) {
      return template;
    }
    return `${prefix}${dir}/${template.slice(idx)}`;
  };
};
function wywInJS({
  debug,
  include,
  exclude,
  sourceMap,
  preserveCssPaths,
  keepComments,
  prefixer,
  preprocessor,
  ssrDevCss,
  ssrDevCssPath,
  transformLibraries,
  ...rest
} = {}) {
  const filter = (0, _vite.createFilter)(include, exclude);
  const cssLookup = {};
  const cssFileLookup = {};
  const pendingCssReloads = new Set();
  let pendingCssReloadTimer;
  let ssrDevCssVersion = 0;
  let config;
  let devServer;
  let importMetaEnvForEval = null;
  const ssrDevCssEnabled = Boolean(ssrDevCss);
  const [ssrDevCssPathname, ssrDevCssQuery] = (ssrDevCssPath !== null && ssrDevCssPath !== void 0 ? ssrDevCssPath : '/_wyw-in-js/ssr.css').split('?', 2);
  const ssrDevCssRoute = ssrDevCssPathname.startsWith('/') ? ssrDevCssPathname : `/${ssrDevCssPathname}`;
  const getSsrDevCssHref = () => {
    const versionParam = `v=${ssrDevCssVersion}`;
    const query = ssrDevCssQuery ? `${ssrDevCssQuery}&${versionParam}` : versionParam;
    return `${ssrDevCssRoute}?${query}`;
  };
  const getSsrDevCssContents = () => {
    const entries = Object.entries(cssLookup);
    if (entries.length === 0) return '';
    const merged = entries.sort(([a], [b]) => a.localeCompare(b)).map(([, cssText]) => cssText).join('\n');
    return `${merged}\n`;
  };
  const {
    emitter,
    onDone
  } = (0, _transform.createFileReporter)(debug !== null && debug !== void 0 ? debug : false);
  const scheduleCssReload = cssFilename => {
    var _devServer;
    if (!((_devServer = devServer) !== null && _devServer !== void 0 && _devServer.moduleGraph)) return;
    pendingCssReloads.add(cssFilename);
    if (pendingCssReloadTimer) return;
    pendingCssReloadTimer = setTimeout(() => {
      pendingCssReloadTimer = undefined;
      const ids = Array.from(pendingCssReloads);
      pendingCssReloads.clear();
      for (const id of ids) {
        const module = devServer.moduleGraph.getModuleById(id);
        if (module) devServer.reloadModule(module);
      }
    }, 0);
  };

  // <dependency id, targets>
  const targets = [];
  const cache = new _transform.TransformCacheCollection();
  const isInsideCacheDir = filename => {
    if (!config.cacheDir) {
      return false;
    }
    const relative = _path.default.relative(config.cacheDir, filename);
    return relative !== '' && !relative.startsWith('..') && !_path.default.isAbsolute(relative);
  };
  const getDepsOptimizer = () => {
    var _ref, _ref2, _server$environments$, _server$environments;
    if (!devServer) return null;
    const server = devServer;
    return (_ref = (_ref2 = (_server$environments$ = (_server$environments = server.environments) === null || _server$environments === void 0 || (_server$environments = _server$environments.client) === null || _server$environments === void 0 ? void 0 : _server$environments.depsOptimizer) !== null && _server$environments$ !== void 0 ? _server$environments$ : server.depsOptimizer) !== null && _ref2 !== void 0 ? _ref2 : server._depsOptimizer) !== null && _ref !== void 0 ? _ref : null;
  };
  const waitForOptimizedDep = async filename => {
    var _depsOptimizer$isOpti, _depsOptimizer$init, _depsOptimizer$metada;
    const depsOptimizer = getDepsOptimizer();
    if (!(depsOptimizer !== null && depsOptimizer !== void 0 && (_depsOptimizer$isOpti = depsOptimizer.isOptimizedDepFile) !== null && _depsOptimizer$isOpti !== void 0 && _depsOptimizer$isOpti.call(depsOptimizer, filename))) {
      return false;
    }
    await ((_depsOptimizer$init = depsOptimizer.init) === null || _depsOptimizer$init === void 0 ? void 0 : _depsOptimizer$init.call(depsOptimizer));
    await depsOptimizer.scanProcessing;
    const info = (_depsOptimizer$metada = depsOptimizer.metadata) === null || _depsOptimizer$metada === void 0 || (_depsOptimizer$metada = _depsOptimizer$metada.depInfoList) === null || _depsOptimizer$metada === void 0 ? void 0 : _depsOptimizer$metada.find(item => item.file === filename);
    if (info !== null && info !== void 0 && info.processing) {
      await info.processing;
    }
    return true;
  };
  const boundResolveCache = new WeakMap();
  const getBoundResolve = ctx => {
    const cached = boundResolveCache.get(ctx);
    if (cached) return cached;
    const boundResolve = (what, importer) => ctx.resolve(what, importer);
    boundResolveCache.set(ctx, boundResolve);
    return boundResolve;
  };
  const createAsyncResolver = (0, _shared.asyncResolverFactory)(async (resolved, what, importer, stack) => {
    const log = _shared.logger.extend('vite').extend((0, _transform.getFileIdx)(importer));
    if (resolved) {
      log("resolve ✅ '%s'@'%s -> %O\n%s", what, importer, resolved);

      // Vite adds param like `?v=667939b3` to cached modules
      let resolvedId = resolved.id.split('?', 1)[0];
      if (resolvedId.startsWith('\0')) {
        // \0 is a special character in Rollup that tells Rollup to not include this in the bundle
        // https://rollupjs.org/guide/en/#outputexports
        return null;
      }
      if (resolvedId.startsWith(VITE_FS_PREFIX)) {
        resolvedId = normalizeViteFsPath(resolvedId);
      }
      if (resolvedId.startsWith('/@')) {
        return null;
      }
      if (resolved.external) {
        // In SSR mode Vite can mark resolved file ids as `external` (including `external: "absolute"`).
        // Prefer Vite's resolved file path and only fall back to Node resolution for bare specifiers.
        if (_path.default.isAbsolute(resolvedId) || (0, _fs.existsSync)(resolvedId)) {
          return resolvedId;
        }

        // If module is marked as external, Rollup will not resolve it further, so resolve it ourselves.
        return (0, _shared.syncResolve)(what, importer, stack);
      }
      if (!(0, _fs.existsSync)(resolvedId)) {
        // When Vite resolves to an optimized deps entry (cacheDir) it may not be written yet.
        // Wait for Vite's optimizer instead of calling optimizeDeps() manually (deprecated in Vite 7).
        try {
          await waitForOptimizedDep(resolvedId);
        } catch {
          // If optimizer failed, fall through to preserve previous behavior and surface the error.
        }

        // Vite can return an optimized deps entry (from cacheDir) before it's written to disk.
        // Manually calling optimizeDeps is deprecated in Vite 7 and can also get called many times.
        // Instead, fall back to resolving the original module path directly.
        if (!(0, _fs.existsSync)(resolvedId) && isInsideCacheDir(resolvedId)) {
          try {
            return (0, _shared.syncResolve)(what, importer, stack);
          } catch {
            // Fall through to preserve previous behavior: return resolvedId and let WyW surface the error.
          }
        }
      }
      return resolvedId;
    }
    log("resolve ❌ '%s'@'%s", what, importer);
    throw new Error(`Could not resolve ${what}`);
  }, (what, importer) => [what, importer]);
  return {
    name: 'wyw-in-js',
    enforce: 'post',
    buildEnd() {
      onDone(process.cwd());
    },
    configResolved(resolvedConfig) {
      var _config$envPrefix;
      config = resolvedConfig;
      if (preserveCssPaths && config.command === 'build') {
        const outputs = config.build.rollupOptions.output;
        let outputEntries = [];
        if (Array.isArray(outputs)) {
          outputEntries = outputs;
        } else if (outputs) {
          outputEntries = [outputs];
        }
        outputEntries.forEach(entry => {
          var _output$assetFileName, _config$build$assetsD;
          if (!entry || typeof entry !== 'object') return;
          const output = entry;
          if (!output.preserveModules) return;
          const template = (_output$assetFileName = output.assetFileNames) !== null && _output$assetFileName !== void 0 ? _output$assetFileName : `${(_config$build$assetsD = config.build.assetsDir) !== null && _config$build$assetsD !== void 0 ? _config$build$assetsD : 'assets'}/[name].[hash].[ext]`;
          const assetFileNames = getWywCssAssetFileNames(config, output, template);
          if (assetFileNames) output.assetFileNames = assetFileNames;
        });
      }
      const envPrefix = (_config$envPrefix = config.envPrefix) !== null && _config$envPrefix !== void 0 ? _config$envPrefix : 'VITE_';
      const envDir =
      // envDir is absolute in modern Vite, but keep a fallback for older versions
      'envDir' in config && typeof config.envDir === 'string' ? config.envDir : config.root;
      const loaded = (0, _vite.loadEnv)(config.mode, envDir, envPrefix);
      const base = {
        ...loaded,
        BASE_URL: config.base,
        MODE: config.mode,
        DEV: config.command === 'serve',
        PROD: config.command === 'build'
      };
      importMetaEnvForEval = {
        client: {
          ...base,
          SSR: false
        },
        ssr: {
          ...base,
          SSR: true
        }
      };
    },
    configureServer(_server) {
      devServer = _server;
      if (!ssrDevCssEnabled || config.command !== 'serve') return;
      devServer.middlewares.use((req, res, next) => {
        const {
          url
        } = req;
        if (!url) {
          next();
          return;
        }
        const [pathname] = url.split('?', 1);
        if (pathname !== ssrDevCssRoute) {
          next();
          return;
        }
        const etag = `W/"${ssrDevCssVersion}"`;
        const ifNoneMatch = req.headers['if-none-match'];
        if (ifNoneMatch === etag) {
          res.statusCode = 304;
          res.end();
          return;
        }
        res.statusCode = 200;
        res.setHeader('Content-Type', 'text/css; charset=utf-8');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('ETag', etag);
        res.end(getSsrDevCssContents());
      });
    },
    transformIndexHtml(html) {
      if (!ssrDevCssEnabled || config.command !== 'serve') return undefined;
      return {
        html,
        tags: [{
          tag: 'link',
          attrs: {
            rel: 'stylesheet',
            href: getSsrDevCssHref()
          },
          injectTo: 'head-prepend'
        }]
      };
    },
    load(url) {
      const [id] = url.split('?', 1);
      return cssLookup[id];
    },
    /* eslint-disable-next-line consistent-return */
    resolveId(importeeUrl) {
      const [id] = importeeUrl.split('?', 1);
      if (cssLookup[id]) return id;
      return cssFileLookup[id];
    },
    handleHotUpdate(ctx) {
      // it's module, so just transform it
      if (ctx.modules.length) return ctx.modules;

      // Select affected modules of changed dependency
      const affected = targets.filter(x =>
      // file is dependency of any target
      x.dependencies.some(dep => dep === ctx.file) ||
      // or changed module is a dependency of any target
      x.dependencies.some(dep => ctx.modules.some(m => m.file === dep)));
      const deps = affected.flatMap(target => target.dependencies);

      // eslint-disable-next-line no-restricted-syntax
      for (const depId of deps) {
        cache.invalidateForFile(depId);
      }
      return affected.map(target => devServer.moduleGraph.getModuleById(target.id)).concat(ctx.modules).filter(m => !!m);
    },
    async transform(code, url, transformOptions) {
      const [id] = url.split('?', 1);

      // Do not transform ignored and generated files
      if (!transformLibraries && url.includes('node_modules') || !filter(url) || id in cssLookup) return;
      const log = _shared.logger.extend('vite').extend((0, _transform.getFileIdx)(id));
      log('transform %s', id);
      const overrideContext = (context, filename) => {
        var _importMetaEnvForEval;
        const isSsr = typeof transformOptions === 'boolean' ? transformOptions : Boolean(transformOptions === null || transformOptions === void 0 ? void 0 : transformOptions.ssr);
        const env = (_importMetaEnvForEval = importMetaEnvForEval) === null || _importMetaEnvForEval === void 0 ? void 0 : _importMetaEnvForEval[isSsr ? 'ssr' : 'client'];
        const withEnv = env ? {
          ...context,
          __wyw_import_meta_env: env
        } : context;
        return rest.overrideContext ? rest.overrideContext(withEnv, filename) : withEnv;
      };
      const transformServices = {
        options: {
          filename: id,
          root: process.cwd(),
          prefixer,
          keepComments,
          preprocessor,
          pluginOptions: {
            ...rest,
            overrideContext
          }
        },
        cache,
        emitWarning: message => this.warn(message),
        eventEmitter: emitter
      };
      const result = await (0, _transform.transform)(transformServices, code, createAsyncResolver(getBoundResolve(this)));
      let {
        cssText,
        dependencies
      } = result;

      // Heads up, there are three cases:
      // 1. cssText is undefined, it means that file was not transformed
      // 2. cssText is empty, it means that file was transformed, but it does not contain any styles
      // 3. cssText is not empty, it means that file was transformed and it contains styles

      if (typeof cssText === 'undefined') {
        return;
      }
      if (cssText === '') {
        /* eslint-disable-next-line consistent-return */
        return {
          code: result.code,
          map: result.sourceMap
        };
      }
      dependencies !== null && dependencies !== void 0 ? dependencies : dependencies = [];
      const cssFilename = _path.default.normalize(`${id.replace(/\.[jt]sx?$/, '')}.wyw-in-js.css`).replace(/\\/g, _path.default.posix.sep);
      const cssRelativePath = _path.default.relative(config.root, cssFilename).replace(/\\/g, _path.default.posix.sep);
      const cssId = `/${cssRelativePath}`;
      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`;
      }
      const didCssChange = cssLookup[cssFilename] !== cssText;
      cssLookup[cssFilename] = cssText;
      cssFileLookup[cssId] = cssFilename;
      result.code += `\nimport ${JSON.stringify(cssFilename)};\n`;
      for (let i = 0, end = dependencies.length; i < end; i++) {
        // eslint-disable-next-line no-await-in-loop
        const depModule = await this.resolve(dependencies[i], url, {
          isEntry: false
        });
        if (depModule) dependencies[i] = depModule.id;
      }
      const target = targets.find(t => t.id === id);
      if (!target) targets.push({
        id,
        dependencies
      });else target.dependencies = dependencies;
      if (didCssChange) {
        scheduleCssReload(cssFilename);
        if (ssrDevCssEnabled && config.command === 'serve') {
          ssrDevCssVersion += 1;
        }
      }
      /* eslint-disable-next-line consistent-return */
      return {
        code: result.code,
        map: result.sourceMap
      };
    }
  };
}
