// @wyw-in-js/transform/types/vm/createVmContext.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVmContext = createVmContext;
const vm = __importStar(require("vm"));
const shared_1 = require("@wyw-in-js/shared");
const process = __importStar(require("./process"));
const NOOP = () => { };
const IMPORT_META_ENV = '__wyw_import_meta_env';
let importMetaEnvWarned = false;
function createImportMetaEnvProxy() {
    const target = Object.create(null);
    const warnOnce = () => {
        if (importMetaEnvWarned)
            return;
        importMetaEnvWarned = true;
        // eslint-disable-next-line no-console
        console.warn([
            `[wyw-in-js] import.meta.env was accessed during build-time evaluation, but no env values were provided.`,
            ``,
            `If you're using Vite, make sure @wyw-in-js/vite plugin is enabled (it injects Vite env for evaluation).`,
            `Otherwise provide "__wyw_import_meta_env" via pluginOptions.overrideContext.`,
        ].join('\n'));
    };
    return new Proxy(target, {
        get(obj, key) {
            if (typeof key === 'symbol') {
                return Reflect.get(obj, key);
            }
            warnOnce();
            return obj[key];
        },
        has(obj, key) {
            if (typeof key === 'symbol') {
                return Reflect.has(obj, key);
            }
            warnOnce();
            return Reflect.has(obj, key);
        },
        getOwnPropertyDescriptor(obj, key) {
            return Reflect.getOwnPropertyDescriptor(obj, key);
        },
        ownKeys(obj) {
            return Reflect.ownKeys(obj);
        },
        set(obj, key, value) {
            if (typeof key === 'symbol') {
                return Reflect.set(obj, key, value);
            }
            warnOnce();
            return Reflect.set(obj, key, value);
        },
    });
}
function createWindow() {
    const { Window, GlobalWindow } = require('happy-dom');
    const HappyWindow = GlobalWindow || Window;
    const win = new HappyWindow();
    // TODO: browser doesn't expose Buffer, but a lot of dependencies use it
    win.Buffer = Buffer;
    win.Uint8Array = Uint8Array;
    return win;
}
/**
 * `happy-dom` already has required references, so we don't need to set them.
 */
function setReferencePropertyIfNotPresent(context, key) {
    if (context[key] === context) {
        return;
    }
    context[key] = context;
}
function createBaseContext(win, additionalContext) {
    const baseContext = win ?? {};
    setReferencePropertyIfNotPresent(baseContext, 'window');
    setReferencePropertyIfNotPresent(baseContext, 'self');
    setReferencePropertyIfNotPresent(baseContext, 'top');
    setReferencePropertyIfNotPresent(baseContext, 'parent');
    setReferencePropertyIfNotPresent(baseContext, 'global');
    setReferencePropertyIfNotPresent(baseContext, 'process');
    baseContext.document = win?.document;
    baseContext.process = process;
    baseContext.clearImmediate = NOOP;
    baseContext.clearInterval = NOOP;
    baseContext.clearTimeout = NOOP;
    baseContext.setImmediate = NOOP;
    baseContext.requestAnimationFrame = NOOP;
    baseContext.setInterval = NOOP;
    baseContext.setTimeout = NOOP;
    // eslint-disable-next-line guard-for-in,no-restricted-syntax
    for (const key in additionalContext) {
        baseContext[key] = additionalContext[key];
    }
    return baseContext;
}
function createHappyDOMWindow() {
    const win = createWindow();
    return {
        teardown: () => {
            win.happyDOM.abort();
        },
        window: win,
    };
}
function createNothing() {
    return {
        teardown: () => { },
        window: undefined,
    };
}
function createVmContext(filename, features, additionalContext, overrideContext = (i) => i) {
    const isHappyDOMEnabled = (0, shared_1.isFeatureEnabled)(features, 'happyDOM', filename);
    const { teardown, window } = isHappyDOMEnabled
        ? createHappyDOMWindow()
        : createNothing();
    const envContext = {
        [IMPORT_META_ENV]: createImportMetaEnvProxy(),
    };
    const baseContext = createBaseContext(window, overrideContext({
        __filename: filename,
        ...envContext,
        ...additionalContext,
    }, filename));
    const context = vm.createContext(baseContext);
    return {
        context,
        teardown,
    };
}
