// @wyw-in-js/transform/lib/cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransformCacheCollection = void 0;
var _crypto = require("crypto");
var _nodeFs = _interopRequireDefault(require("node:fs"));
var _shared = require("@wyw-in-js/shared");
var _getFileIdx = require("./utils/getFileIdx");
var _parseRequest = require("./utils/parseRequest");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function hashContent(content) {
  return (0, _crypto.createHash)('sha256').update(content).digest('hex');
}
const cacheLogger = _shared.logger.extend('cache');
const cacheNames = ['entrypoints', 'exports'];
const loggers = cacheNames.reduce((acc, key) => ({
  ...acc,
  [key]: cacheLogger.extend(key)
}), {});
class TransformCacheCollection {
  contentHashes = new Map();
  constructor(caches = {}) {
    this.entrypoints = caches.entrypoints || new Map();
    this.exports = caches.exports || new Map();
  }
  add(cacheName, key, value) {
    const cache = this[cacheName];
    loggers[cacheName]('%s:add %s %f', (0, _getFileIdx.getFileIdx)(key), key, () => {
      if (value === undefined) {
        return cache.has(key) ? 'removed' : 'noop';
      }
      if (!cache.has(key)) {
        return 'added';
      }
      return cache.get(key) === value ? 'unchanged' : 'updated';
    });
    if (value === undefined) {
      cache.delete(key);
      this.contentHashes.delete(key);
      return;
    }
    cache.set(key, value);
    if ('initialCode' in value) {
      var _value$initialCode;
      this.contentHashes.set(key, hashContent((_value$initialCode = value.initialCode) !== null && _value$initialCode !== void 0 ? _value$initialCode : ''));
    }
  }
  clear(cacheName) {
    if (cacheName === 'all') {
      cacheNames.forEach(name => {
        this.clear(name);
      });
      return;
    }
    loggers[cacheName]('clear');
    const cache = this[cacheName];
    cache.clear();
  }
  delete(cacheName, key) {
    this.invalidate(cacheName, key);
  }
  get(cacheName, key) {
    const cache = this[cacheName];
    const res = cache.get(key);
    loggers[cacheName]('get', key, res === undefined ? 'miss' : 'hit');
    return res;
  }
  has(cacheName, key) {
    const cache = this[cacheName];
    const res = cache.has(key);
    loggers[cacheName]('has', key, res);
    return res;
  }
  invalidate(cacheName, key) {
    const cache = this[cacheName];
    if (!cache.has(key)) {
      return;
    }
    loggers[cacheName]('invalidate', key);
    cache.delete(key);
  }
  invalidateForFile(filename) {
    cacheNames.forEach(cacheName => {
      this.invalidate(cacheName, filename);
    });
  }
  invalidateIfChanged(filename, content, previousVisitedFiles) {
    const visitedFiles = new Set(previousVisitedFiles);
    const fileEntrypoint = this.get('entrypoints', filename);

    // We need to check all dependencies of the file
    // because they might have changed as well.
    if (fileEntrypoint && !visitedFiles.has(filename)) {
      visitedFiles.add(filename);
      for (const [, dependency] of fileEntrypoint.dependencies) {
        const dependencyFilename = dependency.resolved;
        if (dependencyFilename) {
          const dependencyContent = _nodeFs.default.readFileSync((0, _parseRequest.stripQueryAndHash)(dependencyFilename), 'utf8');
          this.invalidateIfChanged(dependencyFilename, dependencyContent, visitedFiles);
        }
      }
    }
    const hash = this.contentHashes.get(filename);
    const newHash = hashContent(content);
    if (hash !== newHash) {
      cacheLogger('content has changed, invalidate all for %s', filename);
      this.contentHashes.set(filename, newHash);
      this.invalidateForFile(filename);
      return true;
    }
    return false;
  }
}
exports.TransformCacheCollection = TransformCacheCollection;
