// @wyw-in-js/transform/lib/module.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Module = exports.DefaultModuleImplementation = void 0;
var _fs = _interopRequireDefault(require("fs"));
var _module = _interopRequireDefault(require("module"));
var _path = _interopRequireDefault(require("path"));
var _vm = _interopRequireDefault(require("vm"));
var _tsInvariant = require("ts-invariant");
var _shared = require("@wyw-in-js/shared");
require("./utils/dispose-polyfill");
var _Entrypoint = require("./transform/Entrypoint");
var _Entrypoint2 = require("./transform/Entrypoint.helpers");
var _UnprocessedEntrypointError = require("./transform/actions/UnprocessedEntrypointError");
var _importOverrides = require("./utils/importOverrides");
var _parseRequest = require("./utils/parseRequest");
var _createVmContext = require("./vm/createVmContext");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * This is a custom implementation for the module system for evaluating code,
 * used for resolving values for dependencies interpolated in `css` or `styled`.
 *
 * This serves 2 purposes:
 * - Avoid leakage from evaluated code to module cache in current context, e.g. `babel-register`
 * - Allow us to invalidate the module cache without affecting other stuff, necessary for rebuilds
 *
 * We also use it to transpile the code with Babel by default.
 * We also store source maps for it to provide correct error stacktraces.
 *
 */

const DefaultModuleImplementation = exports.DefaultModuleImplementation = _module.default;

// Supported node builtins based on the modules polyfilled by webpack
// `true` means module is polyfilled, `false` means module is empty
const builtins = {
  assert: true,
  buffer: true,
  child_process: false,
  cluster: false,
  console: true,
  constants: true,
  crypto: true,
  dgram: false,
  dns: false,
  domain: true,
  events: true,
  fs: false,
  http: true,
  https: true,
  module: false,
  net: false,
  os: true,
  path: true,
  punycode: true,
  process: true,
  querystring: true,
  readline: false,
  repl: false,
  stream: true,
  string_decoder: true,
  sys: true,
  timers: true,
  tls: false,
  tty: true,
  url: true,
  util: true,
  vm: true,
  zlib: true
};
const VITE_VIRTUAL_PREFIX = '/@';
const REACT_REFRESH_VIRTUAL_ID = '/@react-refresh';
const reactRefreshRuntime = {
  createSignatureFunctionForTransform: () => () => {}
};
const NOOP = () => {};
const browserOnlyEvalHintTriggers = ['window is not defined', "evaluating 'window", 'document is not defined', "evaluating 'document", 'navigator is not defined', "evaluating 'navigator", 'self is not defined', "evaluating 'self"];
const getBrowserOnlyEvalHint = error => {
  const message = error instanceof Error ? error.message : String(error);
  const looksLikeBrowserOnly = browserOnlyEvalHintTriggers.some(trigger => message.includes(trigger));
  if (!looksLikeBrowserOnly) return null;
  return ['', '[wyw-in-js] Evaluation hint:', 'This usually means browser-only code ran during build-time evaluation.', 'Move browser-only initialization out of evaluated modules, or mock the import via `importOverrides`.', "Example: importOverrides: { 'msw/browser': { mock: './src/__mocks__/msw-browser.js' } }"].join('\n');
};
const warnedUnknownImportsByServices = new WeakMap();
function emitWarning(services, message) {
  if (services.emitWarning) {
    services.emitWarning(message);
    return;
  }

  // eslint-disable-next-line no-console
  console.warn(message);
}
function getWarnedUnknownImports(services) {
  const cached = warnedUnknownImportsByServices.get(services);
  if (cached) {
    return cached;
  }
  const created = new Set();
  warnedUnknownImportsByServices.set(services, created);
  return created;
}
function getUncached(cached, test) {
  const cachedSet = new Set(typeof cached === 'string' ? cached.split(',') : cached);
  if (cachedSet.has('*')) {
    return [];
  }
  return test.filter(t => !cachedSet.has(t));
}
function resolve(id) {
  const {
    resolved
  } = this.resolveDependency(id);
  (0, _tsInvariant.invariant)(resolved, `Unable to resolve "${id}"`);
  return resolved;
}
const defaultImportLoaders = {
  raw: 'raw',
  url: 'url'
};
class Module {
  callstack = [];
  isEvaluated = false;
  require = Object.assign(id => {
    if (id === REACT_REFRESH_VIRTUAL_ID) {
      this.dependencies.push(id);
      this.debug('require', `vite virtual '${id}'`);
      return reactRefreshRuntime;
    }
    if (id.startsWith(VITE_VIRTUAL_PREFIX)) {
      this.dependencies.push(id);
      this.debug('require', `vite virtual '${id}'`);
      return {};
    }
    if (id in builtins) {
      // The module is in the allowed list of builtin node modules
      // Ideally we should prevent importing them, but webpack polyfills some
      // So we check for the list of polyfills to determine which ones to support
      if (builtins[id]) {
        this.debug('require', `builtin '${id}'`);
        return require(id);
      }
      return null;
    }

    // Resolve module id (and filename) relatively to parent module
    const dependency = this.resolveDependency(id);
    if (dependency.resolved === id && !_path.default.isAbsolute(id)) {
      // The module is a builtin node modules, but not in the allowed list
      throw new Error(`Unable to import "${id}". Importing Node builtins is not supported in the sandbox.`);
    }
    (0, _tsInvariant.invariant)(dependency.resolved, `Dependency ${dependency.source} cannot be resolved`);
    const loaded = this.loadByImportLoaders(id, dependency.resolved);
    if (loaded.handled) {
      this.dependencies.push(id);
      this.debug('require', `${id} -> ${dependency.resolved} (loader)`);
      return loaded.value;
    }
    this.dependencies.push(id);
    this.debug('require', `${id} -> ${dependency.resolved}`);
    const entrypoint = this.getEntrypoint(dependency.resolved, dependency.only, this.debug);
    if (entrypoint === null) {
      return dependency.resolved;
    }
    if (entrypoint.evaluated || (0, _Entrypoint2.isSuperSet)(entrypoint.evaluatedOnly, dependency.only)) {
      return entrypoint.exports;
    }
    const m = this.createChild(entrypoint);
    m.evaluate();
    return entrypoint.exports;
  }, {
    ensure: NOOP,
    resolve: resolve.bind(this)
  });
  resolve = resolve.bind(this);
  #entrypointRef;
  constructor(services, entrypoint, parentModule, moduleImpl = DefaultModuleImplementation) {
    var _parentModule$ignored, _entrypoint$ignored;
    this.services = services;
    this.moduleImpl = moduleImpl;
    this.cache = services.cache;
    this.#entrypointRef = (0, _shared.isFeatureEnabled)(services.options.pluginOptions.features, 'useWeakRefInEval', entrypoint.name) ? new WeakRef(entrypoint) : entrypoint;
    this.idx = entrypoint.idx;
    this.id = entrypoint.name;
    this.filename = entrypoint.name;
    this.dependencies = [];
    this.debug = entrypoint.log.extend('module');
    this.parentIsIgnored = (_parentModule$ignored = parentModule === null || parentModule === void 0 ? void 0 : parentModule.ignored) !== null && _parentModule$ignored !== void 0 ? _parentModule$ignored : false;
    this.ignored = (_entrypoint$ignored = entrypoint.ignored) !== null && _entrypoint$ignored !== void 0 ? _entrypoint$ignored : this.parentIsIgnored;
    if (parentModule) {
      this.callstack = [entrypoint.name, ...parentModule.callstack];
    } else {
      this.callstack = [entrypoint.name];
    }
    this.extensions = services.options.pluginOptions.extensions;
    this.debug('init', entrypoint.name);
  }
  get exports() {
    return this.entrypoint.exports;
  }
  set exports(value) {
    this.entrypoint.exports = value;
    this.debug('the whole exports was overridden with %O', value);
  }
  get entrypoint() {
    const entrypoint = this.#entrypointRef instanceof WeakRef ? this.#entrypointRef.deref() : this.#entrypointRef;
    (0, _tsInvariant.invariant)(entrypoint, `Module ${this.idx} is disposed`);
    return entrypoint;
  }
  evaluate() {
    const {
      entrypoint
    } = this;
    entrypoint.assertTransformed();
    const cached = this.cache.get('entrypoints', entrypoint.name);
    let evaluatedCreated = false;
    if (!entrypoint.supersededWith) {
      this.cache.add('entrypoints', entrypoint.name, entrypoint.createEvaluated());
      evaluatedCreated = true;
    }
    const {
      transformedCode: source
    } = entrypoint;
    const {
      pluginOptions
    } = this.services.options;
    if (!source) {
      this.debug(`evaluate`, 'there is nothing to evaluate');
      return;
    }
    if (this.isEvaluated) {
      this.debug('evaluate', `is already evaluated`);
      return;
    }
    this.debug('evaluate');
    this.debug.extend('source')('%s', source);
    this.isEvaluated = true;
    const filename = (0, _parseRequest.stripQueryAndHash)(this.filename);
    if (/\.json$/.test(filename)) {
      // For JSON files, parse it to a JS object similar to Node
      this.exports = JSON.parse(source);
      return;
    }
    const {
      context,
      teardown
    } = (0, _createVmContext.createVmContext)(filename, pluginOptions.features, {
      module: this,
      exports: entrypoint.exports,
      require: this.require,
      __wyw_dynamic_import: async id => this.require(String(id)),
      __dirname: _path.default.dirname(filename)
    }, pluginOptions.overrideContext);
    try {
      const script = new _vm.default.Script(`(function (exports) { ${source}\n})(exports);`, {
        filename
      });
      script.runInContext(context);
    } catch (e) {
      this.isEvaluated = false;
      if (evaluatedCreated) {
        this.cache.add('entrypoints', entrypoint.name, cached);
      }
      if ((0, _UnprocessedEntrypointError.isUnprocessedEntrypointError)(e)) {
        // It will be handled by evalFile scenario
        throw e;
      }
      if (e instanceof EvalError) {
        this.debug('%O', e);
        throw e;
      }
      this.debug('%O\n%O', e, this.callstack);
      const baseMessage = `${e.message} in${this.callstack.join('\n| ')}\n`;
      const hint = getBrowserOnlyEvalHint(e);
      throw new EvalError(hint ? `${baseMessage}${hint}\n` : baseMessage);
    } finally {
      teardown();
    }
  }
  getEntrypoint(filename, only, log) {
    var _entrypoint$evaluated;
    const strippedFilename = (0, _parseRequest.stripQueryAndHash)(filename);
    const extension = _path.default.extname(strippedFilename);
    if (extension !== '.json' && !this.extensions.includes(extension)) {
      return null;
    }
    const entrypoint = this.cache.get('entrypoints', filename);
    if (entrypoint && (0, _Entrypoint2.isSuperSet)((_entrypoint$evaluated = entrypoint.evaluatedOnly) !== null && _entrypoint$evaluated !== void 0 ? _entrypoint$evaluated : [], only)) {
      log('✅ file has been already evaluated');
      return entrypoint;
    }
    if (entrypoint !== null && entrypoint !== void 0 && entrypoint.ignored) {
      log('✅ file has been ignored during prepare stage. Original code will be used');
      return entrypoint;
    }
    if (this.ignored) {
      log('✅ one of the parent files has been ignored during prepare stage. Original code will be used');
      const newEntrypoint = this.entrypoint.createChild(filename, ['*'], _fs.default.readFileSync(strippedFilename, 'utf-8'));
      if (newEntrypoint === 'loop') {
        const stack = (0, _Entrypoint2.getStack)(this.entrypoint);
        throw new Error(`Circular dependency detected: ${stack.join(' -> ')} -> ${filename}`);
      }
      return newEntrypoint;
    }
    let uncachedExports = null;
    let reprocessOnly = only;
    // Requested file can be already prepared for evaluation on the stage 1
    if (only && entrypoint) {
      var _entrypoint$evaluated2, _entrypoint$only, _entrypoint$evaluated3;
      const evaluatedExports = ((_entrypoint$evaluated2 = entrypoint.evaluatedOnly) === null || _entrypoint$evaluated2 === void 0 ? void 0 : _entrypoint$evaluated2.length) !== 0 ? entrypoint.evaluatedOnly : (_entrypoint$only = entrypoint.only) !== null && _entrypoint$only !== void 0 ? _entrypoint$only : [];
      uncachedExports = getUncached(evaluatedExports, only);
      if (uncachedExports.length === 0) {
        log('✅ ready for evaluation');
        return entrypoint;
      }
      if ((_entrypoint$evaluated3 = entrypoint.evaluatedOnly) !== null && _entrypoint$evaluated3 !== void 0 && _entrypoint$evaluated3.length) {
        reprocessOnly = (0, _Entrypoint2.mergeOnly)(evaluatedExports, only);
      }
      log('❌ file has been processed during prepare stage but %o is not evaluated yet (evaluated: %o)', uncachedExports, evaluatedExports);
    } else {
      log('❌ file has not been processed during prepare stage');
    }

    // If code wasn't extracted from cache, it indicates that we were unable
    // to process some of the imports on stage1. Let's try to reprocess.
    const code = _fs.default.readFileSync(strippedFilename, 'utf-8');
    const newEntrypoint = _Entrypoint.Entrypoint.createRoot(this.services, filename, reprocessOnly, code);
    if (newEntrypoint.evaluated) {
      log('✅ file has been already evaluated');
      return newEntrypoint;
    }
    if (newEntrypoint.ignored) {
      log('✅ file has been ignored during prepare stage. Original code will be used');
      return newEntrypoint;
    }
    return newEntrypoint;
  }
  resolveDependency = id => {
    const cached = this.entrypoint.getDependency(id);
    (0, _tsInvariant.invariant)(!(cached instanceof Promise), 'Dependency is not resolved yet');
    if (cached) {
      return cached;
    }
    if (!this.ignored) {
      this.debug('❌ import has not been resolved during prepare stage. Fallback to Node.js resolver');
    }
    const extensions = this.moduleImpl._extensions;
    const added = [];
    try {
      var _this$services$option, _override$unknown;
      // Check for supported extensions
      this.extensions.forEach(ext => {
        if (ext in extensions) {
          return;
        }

        // When an extension is not supported, add it
        // And keep track of it to clean it up after resolving
        // Use noop for the transform function since we handle it
        extensions[ext] = NOOP;
        added.push(ext);
      });
      const {
        filename
      } = this;
      const strippedId = (0, _parseRequest.stripQueryAndHash)(id);
      let resolved = this.moduleImpl._resolveFilename(strippedId, {
        id: filename,
        filename,
        paths: this.moduleImpl._nodeModulePaths(_path.default.dirname(filename))
      });
      const isFileSpecifier = strippedId.startsWith('.') || _path.default.isAbsolute(strippedId);
      if (isFileSpecifier && _path.default.extname(strippedId) === '' && resolved.endsWith('.cjs') && _fs.default.existsSync(`${resolved.slice(0, -4)}.js`)) {
        // When both `.cjs` and `.js` exist for an extensionless specifier, the
        // resolver may pick `.cjs` depending on the environment/extensions.
        // Prefer `.js` to keep resolved paths stable (e.g. importOverrides keys).
        resolved = `${resolved.slice(0, -4)}.js`;
      }
      const {
        root
      } = this.services.options;
      const keyInfo = (0, _importOverrides.toImportKey)({
        source: id,
        resolved,
        root
      });
      const override = (_this$services$option = this.services.options.pluginOptions.importOverrides) === null || _this$services$option === void 0 ? void 0 : _this$services$option[keyInfo.key];
      const policy = (_override$unknown = override === null || override === void 0 ? void 0 : override.unknown) !== null && _override$unknown !== void 0 ? _override$unknown : override !== null && override !== void 0 && override.mock ? 'allow' : 'warn';
      const shouldWarn = !this.ignored && policy === 'warn';
      let finalResolved = resolved;
      if (override !== null && override !== void 0 && override.mock) {
        try {
          finalResolved = (0, _importOverrides.resolveMockSpecifier)({
            mock: override.mock,
            importer: filename,
            root,
            stack: this.callstack
          });
        } catch (e) {
          var _message;
          const errorMessage = String((_message = e === null || e === void 0 ? void 0 : e.message) !== null && _message !== void 0 ? _message : e);
          throw new Error(`[wyw-in-js] Failed to resolve import mock for "${keyInfo.key}" (${id} from ${filename}): ${errorMessage}`);
        }
      }
      if (policy === 'error') {
        throw new Error([`[wyw-in-js] Unknown import reached during eval (Node resolver fallback)`, ``, `importer: ${filename}`, `source:   ${id}`, `resolved: ${resolved}`, override !== null && override !== void 0 && override.mock ? `mock:     ${override.mock} -> ${finalResolved}` : ``, ``, `callstack:`, ...this.callstack.map(item => `  ${item}`), ``, `config key: ${keyInfo.key}`].filter(Boolean).join('\n'));
      }
      const warnedUnknownImports = getWarnedUnknownImports(this.services);
      if (shouldWarn && !warnedUnknownImports.has(keyInfo.key)) {
        warnedUnknownImports.add(keyInfo.key);
        emitWarning(this.services, [`[wyw-in-js] Unknown import reached during eval (Node resolver fallback)`, ``, `importer: ${filename}`, `source:   ${id}`, `resolved: ${resolved}`, override !== null && override !== void 0 && override.mock ? `mock:     ${override.mock} -> ${finalResolved}` : ``, ``, `callstack:`, ...this.callstack.map(item => `  ${item}`), ``, `config key: ${keyInfo.key}`, `hint: add { importOverrides: { ${JSON.stringify(keyInfo.key)}: { unknown: 'allow' } } } to silence warnings, or use { mock } / { noShake: true } overrides.`].filter(Boolean).join('\n'));
      }
      return {
        source: id,
        only: (0, _importOverrides.applyImportOverrideToOnly)(['*'], override),
        resolved: finalResolved
      };
    } finally {
      // Cleanup the extensions we added to restore previous behaviour
      added.forEach(ext => delete extensions[ext]);
    }
  };
  createChild(entrypoint) {
    return new Module(this.services, entrypoint, this, this.moduleImpl);
  }
  loadByImportLoaders(request, resolved) {
    const {
      pluginOptions
    } = this.services.options;
    const importLoaders = pluginOptions.importLoaders === undefined ? defaultImportLoaders : {
      ...defaultImportLoaders,
      ...pluginOptions.importLoaders
    };
    const {
      query,
      hash
    } = (0, _parseRequest.parseRequest)(request);
    if (!query) return {
      handled: false,
      value: undefined
    };
    const params = new URLSearchParams(query);
    const matchedKey = Array.from(params.keys()).find(key => importLoaders[key] !== undefined && importLoaders[key] !== false);
    if (!matchedKey) return {
      handled: false,
      value: undefined
    };
    const loader = importLoaders[matchedKey];
    const filename = (0, _parseRequest.stripQueryAndHash)(resolved);
    const importer = (0, _parseRequest.stripQueryAndHash)(this.filename);
    const importerDir = _path.default.dirname(importer);
    const toUrl = () => {
      const relative = _path.default.relative(importerDir, filename).replace(/\\/g, _path.default.posix.sep);
      if (relative.startsWith('.') || _path.default.isAbsolute(relative)) {
        return relative;
      }
      return `./${relative}`;
    };
    const readFile = () => _fs.default.readFileSync(filename, 'utf-8');
    const context = {
      importer,
      request,
      resolved,
      filename,
      query,
      hash,
      emitWarning: message => emitWarning(this.services, message),
      readFile,
      toUrl
    };
    if (loader === 'raw') {
      return {
        handled: true,
        value: context.readFile()
      };
    }
    if (loader === 'url') {
      return {
        handled: true,
        value: context.toUrl()
      };
    }
    if (typeof loader === 'function') {
      return {
        handled: true,
        value: loader(context)
      };
    }
    return {
      handled: false,
      value: undefined
    };
  }
}
exports.Module = Module;
