// @wyw-in-js/transform/types/transform/isStaticallyEvaluatableModule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isStaticallyEvaluatableModule = isStaticallyEvaluatableModule;
const types_1 = require("@babel/types");
function isTypeOnlyImport(statement) {
    if (statement.importKind === 'type') {
        return true;
    }
    return statement.specifiers.every((specifier) => (0, types_1.isImportSpecifier)(specifier) && specifier.importKind === 'type');
}
function isTypeOnlyReExport(statement) {
    if (!statement.source) {
        return false;
    }
    return statement.exportKind === 'type';
}
function unwrapExpression(expr) {
    let current = expr;
    for (;;) {
        if ((0, types_1.isTSAsExpression)(current)) {
            current = current.expression;
            // eslint-disable-next-line no-continue
            continue;
        }
        if ((0, types_1.isTSSatisfiesExpression)(current)) {
            current = current.expression;
            // eslint-disable-next-line no-continue
            continue;
        }
        if ((0, types_1.isTSNonNullExpression)(current)) {
            current = current.expression;
            // eslint-disable-next-line no-continue
            continue;
        }
        if ((0, types_1.isTSInstantiationExpression)(current)) {
            current = current.expression;
            // eslint-disable-next-line no-continue
            continue;
        }
        if ((0, types_1.isParenthesizedExpression)(current)) {
            current = current.expression;
            // eslint-disable-next-line no-continue
            continue;
        }
        return current;
    }
}
function isSafeExpression(expr) {
    const unwrapped = unwrapExpression(expr);
    if ((0, types_1.isStringLiteral)(unwrapped) ||
        (0, types_1.isNumericLiteral)(unwrapped) ||
        (0, types_1.isBooleanLiteral)(unwrapped) ||
        (0, types_1.isNullLiteral)(unwrapped) ||
        (0, types_1.isBigIntLiteral)(unwrapped)) {
        return true;
    }
    if ((0, types_1.isArrowFunctionExpression)(unwrapped) || (0, types_1.isFunctionExpression)(unwrapped)) {
        return true;
    }
    if ((0, types_1.isIdentifier)(unwrapped)) {
        return (unwrapped.name === 'undefined' ||
            unwrapped.name === 'NaN' ||
            unwrapped.name === 'Infinity');
    }
    if ((0, types_1.isTemplateLiteral)(unwrapped)) {
        return unwrapped.expressions.every((item) => (0, types_1.isExpression)(item) && isSafeExpression(item));
    }
    if ((0, types_1.isUnaryExpression)(unwrapped)) {
        return isSafeExpression(unwrapped.argument);
    }
    if ((0, types_1.isBinaryExpression)(unwrapped) || (0, types_1.isLogicalExpression)(unwrapped)) {
        return (isSafeExpression(unwrapped.left) &&
            isSafeExpression(unwrapped.right));
    }
    if ((0, types_1.isConditionalExpression)(unwrapped)) {
        return (isSafeExpression(unwrapped.test) &&
            isSafeExpression(unwrapped.consequent) &&
            isSafeExpression(unwrapped.alternate));
    }
    if ((0, types_1.isArrayExpression)(unwrapped)) {
        return unwrapped.elements.every((item) => {
            if (item === null)
                return true;
            if ((0, types_1.isSpreadElement)(item))
                return false;
            return isSafeExpression(item);
        });
    }
    if ((0, types_1.isObjectExpression)(unwrapped)) {
        return unwrapped.properties.every((prop) => {
            if ((0, types_1.isSpreadElement)(prop)) {
                return false;
            }
            if ((0, types_1.isObjectMethod)(prop)) {
                return !prop.computed;
            }
            if ((0, types_1.isObjectProperty)(prop)) {
                if (prop.computed) {
                    return false;
                }
                return (0, types_1.isExpression)(prop.value) && isSafeExpression(prop.value);
            }
            return false;
        });
    }
    if ((0, types_1.isCallExpression)(unwrapped) ||
        (0, types_1.isNewExpression)(unwrapped) ||
        (0, types_1.isTaggedTemplateExpression)(unwrapped) ||
        (0, types_1.isAwaitExpression)(unwrapped) ||
        (0, types_1.isYieldExpression)(unwrapped) ||
        (0, types_1.isUpdateExpression)(unwrapped) ||
        (0, types_1.isAssignmentExpression)(unwrapped) ||
        (0, types_1.isSequenceExpression)(unwrapped)) {
        return false;
    }
    return false;
}
function isSafeDeclarator(declarator) {
    if (!declarator.init) {
        return true;
    }
    return isSafeExpression(declarator.init);
}
function isSafeVariableDeclaration(decl) {
    return decl.declarations.every(isSafeDeclarator);
}
function isSafeStatement(statement) {
    if ((0, types_1.isImportDeclaration)(statement)) {
        return isTypeOnlyImport(statement);
    }
    if ((0, types_1.isExportAllDeclaration)(statement)) {
        return false;
    }
    if ((0, types_1.isExportNamedDeclaration)(statement)) {
        if (!statement.declaration) {
            return !statement.source || isTypeOnlyReExport(statement);
        }
        if ((0, types_1.isFunctionDeclaration)(statement.declaration)) {
            return true;
        }
        if ((0, types_1.isClassDeclaration)(statement.declaration)) {
            return false;
        }
        if (statement.declaration.type === 'VariableDeclaration') {
            return isSafeVariableDeclaration(statement.declaration);
        }
        return false;
    }
    if ((0, types_1.isExportDefaultDeclaration)(statement)) {
        const decl = statement.declaration;
        if ((0, types_1.isFunctionDeclaration)(decl) ||
            (0, types_1.isFunctionExpression)(decl) ||
            (0, types_1.isArrowFunctionExpression)(decl) ||
            (0, types_1.isClassExpression)(decl) ||
            (0, types_1.isClassDeclaration)(decl)) {
            return false;
        }
        return isSafeExpression(decl);
    }
    if (statement.type === 'VariableDeclaration') {
        return isSafeVariableDeclaration(statement);
    }
    if ((0, types_1.isFunctionDeclaration)(statement)) {
        return true;
    }
    if ((0, types_1.isEmptyStatement)(statement)) {
        return true;
    }
    if ((0, types_1.isExpressionStatement)(statement)) {
        // Directives (like "use strict") are safe.
        return (0, types_1.isStringLiteral)(statement.expression);
    }
    return false;
}
function isStaticallyEvaluatableModule(ast) {
    return ast.program.body.every(isSafeStatement);
}
